File: ts/src/tui/theme/resolver.ts
import type { RGBA, ResolvedTheme, ThemeJson, ThemeMode, ThemeValue } from "./types";

export class ResolveError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ResolveError";
  }
}

const HEX_RE = /^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?$/;
const SHORT_HEX_RE = /^#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])?$/;

export function parseHex(hex: string): RGBA {
  const short = hex.match(SHORT_HEX_RE);
  if (short) {
    return {
      r: parseInt(short[1]! + short[1], 16),
      g: parseInt(short[2]! + short[2], 16),
      b: parseInt(short[3]! + short[3], 16),
      a: short[4] ? parseInt(short[4]! + short[4], 16) / 255 : 1,
    };
  }

  const match = hex.match(HEX_RE);
  if (!match) {
    throw new ResolveError(`Invalid hex color: ${hex}`);
  }

  return {
    r: parseInt(match[1]!, 16),
    g: parseInt(match[2]!, 16),
    b: parseInt(match[3]!, 16),
    a: match[4] ? parseInt(match[4]!, 16) / 255 : 1,
  };
}

export function rgbaEqual(a: RGBA, b: RGBA): boolean {
  return a.r === b.r && a.g === b.g && a.b === b.b && a.a === b.a;
}

export function rgbaToString(rgba: RGBA): string {
  if (rgba.a === 1) {
    return `#${rgba.r.toString(16).padStart(2, "0")}${rgba.g.toString(16).padStart(2, "0")}${rgba.b.toString(16).padStart(2, "0")}`;
  }
  return `rgba(${rgba.r},${rgba.g},${rgba.b},${rgba.a})`;
}

export const TRANSPARENT: RGBA = { r: 0, g: 0, b: 0, a: 0 };

function isVariant(value: ThemeValue): value is { dark: string; light: string } {
  return typeof value === "object" && !Array.isArray(value) && "dark" in value && "light" in value;
}

export function resolveColor(
  value: ThemeValue,
  defs: Record<string, string>,
  theme: Record<string, ThemeValue>,
  mode: ThemeMode,
  chain?: string[],
): RGBA {
  const c = chain ?? [];

  if (typeof value === "number") {
    return ansiToRgba(value);
  }

  if (isVariant(value)) {
    return resolveColor(value[mode], defs, theme, mode, c);
  }

  const str = value as string;

  if (str === "transparent" || str === "none") {
    return TRANSPARENT;
  }

  if (str.startsWith("#")) {
    return parseHex(str);
  }

  if (str.startsWith("rgba(") || str.startsWith("rgb(")) {
    return parseCssRgba(str);
  }

  // It's a reference
  if (c.includes(str)) {
    throw new ResolveError(`Circular reference detected: ${c.join(" -> ")} -> ${str}`);
  }

  // Try defs first
  if (defs[str] !== undefined) {
    const defValue = defs[str]!;
    if (defValue.startsWith("#")) {
      return parseHex(defValue);
    }
    if (defValue.startsWith("rgba(") || defValue.startsWith("rgb(")) {
      return parseCssRgba(defValue);
    }
    // Recursive def reference
    return resolveColor(defValue, defs, theme, mode, [...c, str]);
  }

  // Try theme tokens
  const themeValue = theme[str];
  if (themeValue !== undefined) {
    return resolveColor(themeValue, defs, theme, mode, [...c, str]);
  }

  throw new ResolveError(`Missing reference: ${str}`);
}

function parseCssRgba(str: string): RGBA {
  const inner = str.replace(/^rgba?\(/, "").replace(/\)$/, "");
  const parts = inner.split(",").map((p) => p.trim());
  return {
    r: parseInt(parts[0]!, 10),
    g: parseInt(parts[1]!, 10),
    b: parseInt(parts[2]!, 10),
    a: parts[3] ? parseFloat(parts[3]) : 1,
  };
}

export function ansiToRgba(index: number): RGBA {
  // Basic ANSI 16-color mapping
  const ansi16: RGBA[] = [
    { r: 0, g: 0, b: 0, a: 1 },
    { r: 128, g: 0, b: 0, a: 1 },
    { r: 0, g: 128, b: 0, a: 1 },
    { r: 128, g: 128, b: 0, a: 1 },
    { r: 0, g: 0, b: 128, a: 1 },
    { r: 128, g: 0, b: 128, a: 1 },
    { r: 0, g: 128, b: 128, a: 1 },
    { r: 192, g: 192, b: 192, a: 1 },
    { r: 128, g: 128, b: 128, a: 1 },
    { r: 255, g: 0, b: 0, a: 1 },
    { r: 0, g: 255, b: 0, a: 1 },
    { r: 255, g: 255, b: 0, a: 1 },
    { r: 0, g: 0, b: 255, a: 1 },
    { r: 255, g: 0, b: 255, a: 1 },
    { r: 0, g: 255, b: 255, a: 1 },
    { r: 255, g: 255, b: 255, a: 1 },
  ];

  if (index >= 0 && index < 16) {
    return ansi16[index]!;
  }

  if (index >= 16 && index < 232) {
    const i = index - 16;
    const r = (i / 36) % 6;
    const g = (i / 6) % 6;
    const b = i % 6;
    return {
      r: Math.round((r / 5) * 255),
      g: Math.round((g / 5) * 255),
      b: Math.round((b / 5) * 255),
      a: 1,
    };
  }

  if (index >= 232 && index < 256) {
    const gray = Math.round(((index - 232) / 23) * 255);
    return { r: gray, g: gray, b: gray, a: 1 };
  }

  return { r: 0, g: 0, b: 0, a: 1 };
}

function resolveOptional(
  key: string,
  defs: Record<string, string>,
  theme: Record<string, ThemeValue>,
  mode: ThemeMode,
): RGBA | undefined {
  const value = theme[key];
  if (value === undefined) return undefined;
  try {
    return resolveColor(value, defs, theme, mode);
  } catch {
    return undefined;
  }
}

export function resolveTheme(themeJson: ThemeJson, mode: ThemeMode): ResolvedTheme {
  const { defs = {}, theme } = themeJson;

  const r = (key: string): RGBA => {
    const value = theme[key];
    if (value === undefined) {
      throw new ResolveError(`Missing theme token: ${key}`);
    }
    return resolveColor(value, defs, theme, mode);
  };

  const resolved: ResolvedTheme = {
    primary: r("primary"),
    secondary: r("secondary"),
    accent: r("accent"),
    error: r("error"),
    warning: r("warning"),
    success: r("success"),
    info: r("info"),
    text: r("text"),
    textMuted: r("textMuted"),
    selectedListItemText: resolveOptional("selectedListItemText", defs, theme, mode) ?? r("background"),
    background: r("background"),
    backgroundPanel: r("backgroundPanel"),
    backgroundElement: r("backgroundElement"),
    backgroundMenu: resolveOptional("backgroundMenu", defs, theme, mode) ?? r("backgroundElement"),
    border: r("border"),
    borderActive: r("borderActive"),
    borderSubtle: r("borderSubtle"),
    diffAdded: r("diffAdded"),
    diffRemoved: r("diffRemoved"),
    diffContext: r("diffContext"),
    diffHunkHeader: r("diffHunkHeader"),
    diffHighlightAdded: r("diffHighlightAdded"),
    diffHighlightRemoved: r("diffHighlightRemoved"),
    diffAddedBg: r("diffAddedBg"),
    diffRemovedBg: r("diffRemovedBg"),
    diffContextBg: r("diffContextBg"),
    diffLineNumber: r("diffLineNumber"),
    diffAddedLineNumberBg: r("diffAddedLineNumberBg"),
    diffRemovedLineNumberBg: r("diffRemovedLineNumberBg"),
    markdownText: r("markdownText"),
    markdownHeading: r("markdownHeading"),
    markdownLink: r("markdownLink"),
    markdownLinkText: r("markdownLinkText"),
    markdownCode: r("markdownCode"),
    markdownBlockQuote: r("markdownBlockQuote"),
    markdownEmph: r("markdownEmph"),
    markdownStrong: r("markdownStrong"),
    markdownHorizontalRule: r("markdownHorizontalRule"),
    markdownListItem: r("markdownListItem"),
    markdownListEnumeration: r("markdownListEnumeration"),
    markdownImage: r("markdownImage"),
    markdownImageText: r("markdownImageText"),
    markdownCodeBlock: r("markdownCodeBlock"),
    syntaxComment: r("syntaxComment"),
    syntaxKeyword: r("syntaxKeyword"),
    syntaxFunction: r("syntaxFunction"),
    syntaxVariable: r("syntaxVariable"),
    syntaxString: r("syntaxString"),
    syntaxNumber: r("syntaxNumber"),
    syntaxType: r("syntaxType"),
    syntaxOperator: r("syntaxOperator"),
    syntaxPunctuation: r("syntaxPunctuation"),
    thinkingOpacity: typeof theme.thinkingOpacity === "number" ? theme.thinkingOpacity : 0.6,
  };

  return resolved;
}

export function resolveThemeSafe(themeJson: ThemeJson, mode: ThemeMode): ResolvedTheme | null {
  try {
    return resolveTheme(themeJson, mode);
  } catch {
    return null;
  }
}

File: crates/frost-core/src/theme/resolver.rs
use crate::theme::types::{ResolvedTheme, RGBA, ThemeDefValue, ThemeJson, ThemeMode, ThemeValue};
use thiserror::Error;

#[derive(Debug, Error)]
#[error("Theme resolve error: {0}")]
pub struct ResolveError(pub String);

const HEX_RE: &str = r"^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?$";
const SHORT_HEX_RE: &str = r"^#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])?$";

pub fn parse_hex(hex: &str) -> Result<RGBA, ResolveError> {
    let short_re = regex_lite::Regex::new(SHORT_HEX_RE).unwrap();
    if let Some(caps) = short_re.captures(hex) {
        let r = u8::from_str_radix(&format!("{}{}", &caps[1], &caps[1]), 16).unwrap();
        let g = u8::from_str_radix(&format!("{}{}", &caps[2], &caps[2]), 16).unwrap();
        let b = u8::from_str_radix(&format!("{}{}", &caps[3], &caps[3]), 16).unwrap();
        let a = if let Some(a_cap) = caps.get(4) {
            u8::from_str_radix(&format!("{}{}", a_cap.as_str(), a_cap.as_str()), 16).unwrap() as f32
                / 255.0
        } else {
            1.0
        };
        return Ok(RGBA::from_u8(r, g, b, a));
    }

    let long_re = regex_lite::Regex::new(HEX_RE).unwrap();
    let caps = long_re
        .captures(hex)
        .ok_or_else(|| ResolveError(format!("Invalid hex color: {}", hex)))?;

    let r = u8::from_str_radix(&caps[1], 16).unwrap();
    let g = u8::from_str_radix(&caps[2], 16).unwrap();
    let b = u8::from_str_radix(&caps[3], 16).unwrap();
    let a = if let Some(a_cap) = caps.get(4) {
        u8::from_str_radix(a_cap.as_str(), 16).unwrap() as f32 / 255.0
    } else {
        1.0
    };

    Ok(RGBA::from_u8(r, g, b, a))
}

fn parse_css_rgba(str: &str) -> Result<RGBA, ResolveError> {
    let inner = str
        .trim_start_matches("rgba(")
        .trim_start_matches("rgb(")
        .trim_end_matches(')');
    let parts: Vec<&str> = inner.split(',').map(|s| s.trim()).collect();
    if parts.len() < 3 {
        return Err(ResolveError(format!("Invalid rgb/rgba: {}", str)));
    }
    let r = parts[0].parse::<u8>().map_err(|e| ResolveError(e.to_string()))?;
    let g = parts[1].parse::<u8>().map_err(|e| ResolveError(e.to_string()))?;
    let b = parts[2].parse::<u8>().map_err(|e| ResolveError(e.to_string()))?;
    let a = parts.get(3).and_then(|s| s.parse::<f32>().ok()).unwrap_or(1.0);
    Ok(RGBA::from_u8(r, g, b, a))
}

pub fn ansi_to_rgba(index: u8) -> RGBA {
    let ansi16: [RGBA; 16] = [
        RGBA::from_u8(0, 0, 0, 1.0),
        RGBA::from_u8(128, 0, 0, 1.0),
        RGBA::from_u8(0, 128, 0, 1.0),
        RGBA::from_u8(128, 128, 0, 1.0),
        RGBA::from_u8(0, 0, 128, 1.0),
        RGBA::from_u8(128, 0, 128, 1.0),
        RGBA::from_u8(0, 128, 128, 1.0),
        RGBA::from_u8(192, 192, 192, 1.0),
        RGBA::from_u8(128, 128, 128, 1.0),
        RGBA::from_u8(255, 0, 0, 1.0),
        RGBA::from_u8(0, 255, 0, 1.0),
        RGBA::from_u8(255, 255, 0, 1.0),
        RGBA::from_u8(0, 0, 255, 1.0),
        RGBA::from_u8(255, 0, 255, 1.0),
        RGBA::from_u8(0, 255, 255, 1.0),
        RGBA::from_u8(255, 255, 255, 1.0),
    ];

    if index < 16 {
        return ansi16[index as usize];
    }

    if index < 232 {
        let i = index - 16;
        let r = (i / 36) % 6;
        let g = (i / 6) % 6;
        let b = i % 6;
        return RGBA::from_u8(
            ((r as f32 / 5.0) * 255.0).round() as u8,
            ((g as f32 / 5.0) * 255.0).round() as u8,
            ((b as f32 / 5.0) * 255.0).round() as u8,
            1.0,
        );
    }

    let gray = ((index - 232) as f32 / 23.0 * 255.0).round() as u8;
    RGBA::from_u8(gray, gray, gray, 1.0)
}

pub fn resolve_color(
    value: &ThemeValue,
    defs: &std::collections::HashMap<String, ThemeDefValue>,
    theme: &std::collections::HashMap<String, ThemeValue>,
    mode: ThemeMode,
    chain: &mut Vec<String>,
) -> Result<RGBA, ResolveError> {
    match value {
        ThemeValue::Number(n) => Ok(ansi_to_rgba(*n as u8)),
        ThemeValue::Variant { dark, light } => {
            let s = match mode {
                ThemeMode::Dark => dark,
                ThemeMode::Light => light,
            };
            resolve_color(&ThemeValue::String(s.clone()), defs, theme, mode, chain)
        }
        ThemeValue::String(str) => {
            if str == "transparent" || str == "none" {
                return Ok(RGBA::TRANSPARENT);
            }
            if str.starts_with('#') {
                return parse_hex(str);
            }
            if str.starts_with("rgba(") || str.starts_with("rgb(") {
                return parse_css_rgba(str);
            }

            // Reference resolution
            if chain.contains(str) {
                return Err(ResolveError(format!(
                    "Circular reference detected: {} -> {}",
                    chain.join(" -> "),
                    str
                )));
            }

            // Try defs first
            if let Some(def_value) = defs.get(str) {
                let def_str = def_value.as_string();
                if def_str.starts_with('#') {
                    return parse_hex(&def_str);
                }
                if def_str.starts_with("rgba(") || def_str.starts_with("rgb(") {
                    return parse_css_rgba(&def_str);
                }
                chain.push(str.clone());
                let result = resolve_color(
                    &ThemeValue::String(def_str),
                    defs,
                    theme,
                    mode,
                    chain,
                );
                chain.pop();
                return result;
            }

            // Try theme tokens
            if let Some(theme_value) = theme.get(str) {
                chain.push(str.clone());
                let result = resolve_color(theme_value, defs, theme, mode, chain);
                chain.pop();
                return result;
            }

            Err(ResolveError(format!("Missing reference: {}", str)))
        }
    }
}

fn resolve_optional(
    key: &str,
    defs: &std::collections::HashMap<String, ThemeDefValue>,
    theme: &std::collections::HashMap<String, ThemeValue>,
    mode: ThemeMode,
) -> Option<RGBA> {
    let value = theme.get(key)?;
    let mut chain = Vec::new();
    resolve_color(value, defs, theme, mode, &mut chain).ok()
}

pub fn resolve_theme(theme_json: &ThemeJson, mode: ThemeMode) -> Result<ResolvedTheme, ResolveError> {
    let defs = theme_json.defs.clone().unwrap_or_default();
    let theme = &theme_json.theme;

    let r = |key: &str| -> Result<RGBA, ResolveError> {
        let value = theme.get(key).ok_or_else(|| {
            ResolveError(format!("Missing theme token: {}", key))
        })?;
        let mut chain = Vec::new();
        resolve_color(value, &defs, theme, mode, &mut chain)
    };

    Ok(ResolvedTheme {
        primary: r("primary")?,
        secondary: r("secondary")?,
        accent: r("accent")?,
        error: r("error")?,
        warning: r("warning")?,
        success: r("success")?,
        info: r("info")?,
        text: r("text")?,
        text_muted: r("textMuted")?,
        selected_list_item_text: resolve_optional("selectedListItemText", &defs, theme, mode)
            .unwrap_or_else(|| r("background").unwrap_or(RGBA::TRANSPARENT)),
        background: r("background")?,
        background_panel: r("backgroundPanel")?,
        background_element: r("backgroundElement")?,
        background_menu: resolve_optional("backgroundMenu", &defs, theme, mode)
            .unwrap_or_else(|| r("backgroundElement").unwrap_or(RGBA::TRANSPARENT)),
        border: r("border")?,
        border_active: r("borderActive")?,
        border_subtle: r("borderSubtle")?,
        diff_added: r("diffAdded")?,
        diff_removed: r("diffRemoved")?,
        diff_context: r("diffContext")?,
        diff_hunk_header: r("diffHunkHeader")?,
        diff_highlight_added: r("diffHighlightAdded")?,
        diff_highlight_removed: r("diffHighlightRemoved")?,
        diff_added_bg: r("diffAddedBg")?,
        diff_removed_bg: r("diffRemovedBg")?,
        diff_context_bg: r("diffContextBg")?,
        diff_line_number: r("diffLineNumber")?,
        diff_added_line_number_bg: r("diffAddedLineNumberBg")?,
        diff_removed_line_number_bg: r("diffRemovedLineNumberBg")?,
        markdown_text: r("markdownText")?,
        markdown_heading: r("markdownHeading")?,
        markdown_link: r("markdownLink")?,
        markdown_link_text: r("markdownLinkText")?,
        markdown_code: r("markdownCode")?,
        markdown_block_quote: r("markdownBlockQuote")?,
        markdown_emph: r("markdownEmph")?,
        markdown_strong: r("markdownStrong")?,
        markdown_horizontal_rule: r("markdownHorizontalRule")?,
        markdown_list_item: r("markdownListItem")?,
        markdown_list_enumeration: r("markdownListEnumeration")?,
        markdown_image: r("markdownImage")?,
        markdown_image_text: r("markdownImageText")?,
        markdown_code_block: r("markdownCodeBlock")?,
        syntax_comment: r("syntaxComment")?,
        syntax_keyword: r("syntaxKeyword")?,
        syntax_function: r("syntaxFunction")?,
        syntax_variable: r("syntaxVariable")?,
        syntax_string: r("syntaxString")?,
        syntax_number: r("syntaxNumber")?,
        syntax_type: r("syntaxType")?,
        syntax_operator: r("syntaxOperator")?,
        syntax_punctuation: r("syntaxPunctuation")?,
        thinking_opacity: theme
            .get("thinkingOpacity")
            .and_then(|v| match v {
                ThemeValue::Number(n) => Some(*n),
                _ => None,
            })
            .unwrap_or(0.6),
    })
}

pub fn resolve_theme_safe(theme_json: &ThemeJson, mode: ThemeMode) -> Option<ResolvedTheme> {
    resolve_theme(theme_json, mode).ok()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn test_parse_hex() {
        assert_eq!(parse_hex("#ff0000").unwrap(), RGBA::from_u8(255, 0, 0, 1.0));
        assert_eq!(parse_hex("#00ff00").unwrap(), RGBA::from_u8(0, 255, 0, 1.0));
        assert_eq!(parse_hex("#0000ff").unwrap(), RGBA::from_u8(0, 0, 255, 1.0));
        assert_eq!(parse_hex("#fff").unwrap(), RGBA::from_u8(255, 255, 255, 1.0));
        assert_eq!(parse_hex("#f00").unwrap(), RGBA::from_u8(255, 0, 0, 1.0));
        assert_eq!(parse_hex("#ff000080").unwrap(), RGBA::from_u8(255, 0, 0, 128.0 / 255.0));
    }

    #[test]
    fn test_ansi_to_rgba() {
        assert_eq!(ansi_to_rgba(0), RGBA::from_u8(0, 0, 0, 1.0));
        assert_eq!(ansi_to_rgba(15), RGBA::from_u8(255, 255, 255, 1.0));
    }

    #[test]
    fn test_resolve_color_basic() {
        let defs = HashMap::new();
        let theme = HashMap::new();
        let mut chain = Vec::new();

        let value = ThemeValue::String("#ff0000".to_string());
        assert_eq!(
            resolve_color(&value, &defs, &theme, ThemeMode::Dark, &mut chain).unwrap(),
            RGBA::from_u8(255, 0, 0, 1.0)
        );
    }

    #[test]
    fn test_resolve_color_variant() {
        let defs = HashMap::new();
        let theme = HashMap::new();
        let mut chain = Vec::new();

        let value = ThemeValue::Variant {
            dark: "#000000".to_string(),
            light: "#ffffff".to_string(),
        };
        assert_eq!(
            resolve_color(&value, &defs, &theme, ThemeMode::Dark, &mut chain).unwrap(),
            RGBA::from_u8(0, 0, 0, 1.0)
        );
        assert_eq!(
            resolve_color(&value, &defs, &theme, ThemeMode::Light, &mut chain).unwrap(),
            RGBA::from_u8(255, 255, 255, 1.0)
        );
    }

    #[test]
    fn test_resolve_color_reference() {
        let mut defs = HashMap::new();
        defs.insert("myRed".to_string(), ThemeDefValue::String("#ff0000".to_string()));
        let theme = HashMap::new();
        let mut chain = Vec::new();

        let value = ThemeValue::String("myRed".to_string());
        assert_eq!(
            resolve_color(&value, &defs, &theme, ThemeMode::Dark, &mut chain).unwrap(),
            RGBA::from_u8(255, 0, 0, 1.0)
        );
    }

    #[test]
    fn test_resolve_theme_opencode() {
        let theme_json = crate::theme::builtin::builtin_themes()
            .remove("opencode")
            .unwrap();
        let resolved = resolve_theme(&theme_json, ThemeMode::Dark).unwrap();
        assert!(resolved.primary.r > 0.0);
    }
}
